packages/frontend/app/routes/webhooks.tsx

```diff
@@ -0,0 +1,164 @@
+import { json, type LoaderArgs } from '@remix-run/node'
+import {
+  Outlet,
+  useLoaderData,
+  useNavigate,
+  useSearchParams
+} from '@remix-run/react'
+import { PageHeader } from '~/components'
+import { PopoverFilter } from '~/components/Filters'
+import { Button, Table } from '~/components/ui'
+import { listWebhooks } from '~/lib/api/webhook.server'
+import { webhooksSearchParams } from '~/lib/validate.server'
+import { WebhookEventType } from '~/shared/enums'
+
+export const loader = async ({ request }: LoaderArgs) => {
+  const url = new URL(request.url)
+  const searchParams = Object.fromEntries(url.searchParams.entries())
+  const result = webhooksSearchParams.safeParse(
+    searchParams.type
+      ? { ...searchParams, type: searchParams.type.split(',') }
+      : searchParams
+  )
+
+  if (!result.success) {
+    throw json(null, { status: 400, statusText: 'Invalid result.' })
+  }
+
+  const { type, ...pagination } = result.data
+  const webhooks = await listWebhooks({
+    ...pagination,
+    ...(type ? { filter: { type: { in: type } } } : {})
+  })
+
+  let previousPageUrl = '',
+    nextPageUrl = ''
+
+  if (webhooks.pageInfo.hasPreviousPage) {
+    previousPageUrl = `${webhooks.pageInfo.startCursor}`
+  }
+
+  if (webhooks.pageInfo.hasNextPage) {
+    nextPageUrl = `${webhooks.pageInfo.endCursor}`
+  }
+
+  return json({ webhooks, previousPageUrl, nextPageUrl, type })
+}
+
+export default function WebhookEventsPage() {
+  const { webhooks, previousPageUrl, nextPageUrl, type } =
+    useLoaderData<typeof loader>()
+  const [searchParams, setSearchParams] = useSearchParams()
+  const navigate = useNavigate()
+  function setTypeFilterParams(selectedType: WebhookEventType): void {
+    // When modifying a filter load the first page of results
+    searchParams.delete('before')
+    searchParams.delete('after')
+    const selectedTypeSet: Set<WebhookEventType> = type
+      ? new Set(type)
+      : new Set<WebhookEventType>()
+    selectedTypeSet.has(selectedType)
+      ? selectedTypeSet.delete(selectedType)
+      : selectedTypeSet.add(selectedType)
+    if (selectedTypeSet.size > 0) {
+      searchParams.set('type', Array.from(selectedTypeSet).join(','))
+    } else {
+      searchParams.delete('type')
+    }
+    setSearchParams(searchParams)
+  }
```

Comment:
I don't really know what to say about using a `Set` for this. 🤔 Let me know what you think about this approach:
```suggestion
  const [, setSearchParams] = useSearchParams()
  const navigate = useNavigate()

  function setTypeFilterParams(selectedType: WebhookEventType): void {
    const selected = type

    if (!selected.includes(selectedType)) {
      selected.push(selectedType)
      setSearchParams(new URLSearchParams({ type: selected.join(',') }))
      return
    }

    setSearchParams(() => {
      const newParams = selected.filter((t) => t !== selectedType).join(',')
      if (newParams.length === 0) {
        return ''
      }

      return new URLSearchParams({
        type: newParams
      })
    })
  }
```
```

Reply:
Yes, that's much better, thanks.